src/renderer/modals/AddAccounts/steps/StepChooseCurrency.js

```diff
@@ -54,25 +57,97 @@ const StepChooseCurrency = ({ currency, setCurrency }: StepProps) => {
   );
 };
 
-export const StepChooseCurrencyFooter = ({ transitionTo, currency }: StepProps) => {
+export const StepChooseCurrencyFooter = ({
+  transitionTo,
+  currency,
+  existingAccounts,
+  onCloseModal,
+  setCurrency,
+}: StepProps) => {
   const { t } = useTranslation();
-  const url =
-    currency && currency.type === "TokenCurrency"
-      ? supportLinkByTokenType[currency.tokenType]
-      : null;
+  const dispatch = useDispatch();
+  const flatAccounts = flattenAccounts(existingAccounts);
+  const isToken = currency && currency.type === "TokenCurrency";
+
+  // $FlowFixMe
+  const url = isToken ? supportLinkByTokenType[currency.tokenType] : null;
+
+  // $FlowFixMe
+  const parentCurrency = isToken && currency.parentCurrency;
+
+  const parentTokenAccount =
+    isToken &&
+    parentCurrency &&
+    flatAccounts.find(acc => {
+      const c = getAccountCurrency(acc);
+      return c.id === parentCurrency.id;
+    });
+
+  const tokenAccount =
+    isToken &&
+    flatAccounts.find(acc => {
+      const c = getAccountCurrency(acc);
+      // $FlowFixMe
+      return c.id === currency.id;
+    });
+
+  // specific cta in case of token accounts
+  const onTokenCta = useCallback(() => {
+    if (parentTokenAccount) {
+      onCloseModal();
+      dispatch(
+        openModal(
+          "MODAL_RECEIVE",
+          tokenAccount // if already has token receive directly to it
+            ? {
+                account: tokenAccount,
+                parentAccount: parentTokenAccount,
```

Comment:
by the presence of this `{account,parentAccount}` pair, I have the feeling we assume that the parentAccount is account of account,
but it is not necessarily the case look at the current logic you have above

```
  const parentTokenAccount =
    isToken &&
    parentCurrency &&
    flatAccounts.find(acc => {
      const c = getAccountCurrency(acc);
      return c.id === parentCurrency.id;
    });

  const tokenAccount =
    isToken &&
    flatAccounts.find(acc => {
      const c = getAccountCurrency(acc);
      // $FlowFixMe
      return c.id === currency.id;
    });
```

we seem to take the first match of the current and token but independently so it will pick the first, regardless of it one belong to the other.

My second feedback regarding this is I think we can be much more optimized and minimal if we just implement it like this:


```
function findFirstTokenAccount (token, accounts) {
  for (const parentAccount of accounts) {
    for (const account of accounts.subAccounts) {
      if (...) {
        return { account, parentAccount } 
      }
    }
  }
}
```

function that would make a lot of sense to be in live-common.

There are a lot of places where we tend to prefer using the flattenAccounts but it is often minimal and better to stick to the nested version.

(pseudo code, missing some part, but you get the idea)

Reply:
I updated it accordingly i'll create a PR on live-common with this function :+1: